src/resolver/manifest.cc

```diff
@@ -0,0 +1,533 @@
+to_json( nlohmann::json & jto, const ManifestRaw::Options & opts )
+{
+  if ( opts.systems.has_value() ) { jto = { { "systems", *opts.systems } }; }
+  else { jto = nlohmann::json::object(); }
+
+  if ( opts.allow.has_value() ) { jto.emplace( "allow", *opts.allow ); }
+
+  if ( opts.semver.has_value() ) { jto.emplace( "semver", *opts.semver ); }
+
+  if ( opts.packageGroupingStrategy.has_value() )
+    {
+      jto.emplace( "package-grouping-strategy", *opts.packageGroupingStrategy );
+    }
+
+  if ( opts.activationStrategy.has_value() )
+    {
+      jto.emplace( "activation-strategy", *opts.activationStrategy );
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+static void
+from_json( const nlohmann::json & jfrom, ManifestRaw::Hook & hook )
+{
+  if ( ! jfrom.is_object() )
+    {
+      std::string aOrAn = jfrom.is_array() ? " an " : " a ";
+      throw InvalidManifestFileException(
+        "Manifest field `options' must be an object, but is" + aOrAn
+        + std::string( jfrom.type_name() ) + '.' );
+    }
+
+  /* Clear fields. */
+  hook.file   = std::nullopt;
+  hook.script = std::nullopt;
+
+  for ( const auto & [key, value] : jfrom.items() )
+    {
+      if ( key == "script" )
+        {
+          try
+            {
+              value.get_to( hook.script );
+            }
+          catch ( const nlohmann::json::exception & )
+            {
+              throw InvalidManifestFileException(
+                "Failed to parse manifest field `hook.script' with value: "
+                + value.dump() );
+            }
+        }
+      else if ( key == "file" )
+        {
+          try
+            {
+              value.get_to( hook.file );
+            }
+          catch ( const nlohmann::json::exception & )
+            {
+              throw InvalidManifestFileException(
+                "Failed to parse manifest field `hook.file' with value: "
+                + value.dump() );
+            }
+        }
+      else
+        {
+          throw InvalidManifestFileException(
+            "Unrecognized manifest field `hook." + key + "'." );
+        }
+    }
+
+  if ( auto err = hook.check(); err.has_value() ) { throw *err; }
+}
+
+
+[[maybe_unused]] static void
+to_json( nlohmann::json & jto, const ManifestRaw::Hook & hook )
+{
+  if ( auto err = hook.check(); err.has_value() ) { throw *err; }
+  if ( hook.file.has_value() ) { jto = { { "file", *hook.file } }; }
+  else if ( hook.script.has_value() ) { jto = { { "script", *hook.script } }; }
+  else { jto = nlohmann::json::object(); }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+std::optional<InvalidManifestFileException>
+ManifestRaw::Hook::check() const
+{
+  if ( this->script.has_value() && this->file.has_value() )
+    {
+      return InvalidManifestFileException(
+        "Hook may only define one of `hook.script' or `hook.file' fields." );
+    }
+  return std::nullopt;
```

Comment:
suggestion/question nonblocking: should this check that one of them has a value? Since hook is optional, why would we ever want to have a hook with two optional fields?

Reply:
I believe the field is optional.